include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -34,25 +34,70 @@ namespace __par_backend_hetero
+                           const _TransformPattern __transform_pattern, const _ReducePattern __reduce_pattern,
+                           const _InitType __init, _AccLocal& __local_mem, _Res& __res_acc, const _Acc&... __acc)
+{
+    auto __local_idx = __item_id.get_local_id(0);
+    // 1. Initialization (transform part). Fill local memory
+    __transform_pattern(__item_id, __n, /*global_offset*/ 0, __local_mem, __acc...);
+    __dpl_sycl::__group_barrier(__item_id);
+    // 2. Reduce within work group using local memory
+    _Tp __result = __reduce_pattern(__item_id, __n_items, __local_mem);
+    if (__local_idx == 0)
+    {
+        __reduce_pattern.apply_init(__init, __result);
+        __res_acc[0] = __result;
+    }
+}
+
+// Device kernel that transforms and reduces __n elements to the number of work groups preliminary results.
+template <typename _Tp, typename _NDItemId, typename _Size, typename _TransformPattern, typename _ReducePattern,
+          typename _InitType, typename _AccLocal, typename _Tmp, typename... _Acc>
+void
+__device_reduce_kernel(const _NDItemId __item_id, const _Size __n, const _Size __n_items,
+                       const _TransformPattern __transform_pattern, const _ReducePattern __reduce_pattern,
+                       const _InitType __init, _AccLocal& __local_mem, _Tmp& __temp_acc, const _Acc&... __acc)
+{
+    auto __local_idx = __item_id.get_local_id(0);
+    auto __group_idx = __item_id.get_group(0);
+    // 1. Initialization (transform part). Fill local memory
+    __transform_pattern(__item_id, __n, /*global_offset*/ 0, __local_mem, __acc...);
+    __dpl_sycl::__group_barrier(__item_id);
+    // 2. Reduce within work group using local memory
+    _Tp __result = __reduce_pattern(__item_id, __n_items, __local_mem);
+    if (__local_idx == 0)
+        __temp_acc[__group_idx] = __result;
+}
+
 //------------------------------------------------------------------------
 // parallel_transform_reduce - async patterns
 // Please see the comment for __parallel_for_submitter for optional kernel name explanation
 //------------------------------------------------------------------------
 
-// Sequential parallel_transform_reduce used for small input sizes
-template <typename _Tp, typename _KernelName>
-struct __parallel_transform_reduce_seq_submitter;
+// Parallel_transform_reduce for a small arrays using a single work group.
+// Transforms and reduces __work_group_size * __iters_per_work_item elements.
+template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename _KernelName>
+struct __parallel_transform_reduce_small_submitter;
 
-template <typename _Tp, typename... _Name>
-struct __parallel_transform_reduce_seq_submitter<_Tp, __internal::__optional_kernel_name<_Name...>>
+template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename... _Name>
```

Comment:
Since the input size is small here, can `__iters_per_work_item` be a type smaller than `std::size_t`?